### PR TITLE
Documenting the difference betwen df[!, :col] and df[:, :col]

### PR DESCRIPTION
**Pull request by aminya**
Fixes https://github.com/JuliaData/DataFrames.jl/issues/1999

### CODE REVIEW

File: docs/src/man/getting_started.md [hunk 1]
```diff
@@ -45,7 +45,9 @@ julia> df = DataFrame(A = 1:4, B = ["M", "F", "F", "M"])
 
 ```
 
-Columns can be directly (i.e. without copying) accessed via `df.col` or `df[!, :col]`. The latter syntax is more flexible as it allows passing a variable holding the name of the column, and not only a literal name. Note that column names are symbols (`:col` or `Symbol("col")`) rather than strings (`"col"`). Columns can also be accessed using an integer index specifying their position.
+Columns can be directly (i.e. without copying) accessed via `df.col` or `df[!, :col]`. The latter syntax is more flexible as it allows passing a variable holding the name of the column, and not only a literal name. Note that column names are symbols (`:col` or `Symbol("col")`) rather than strings (`"col"`). Columns can also be accessed using an integer index specifying their position. 
+
+`Notice that, changing the elements of the array returned by `df[!, :col]` will affect the original `df`, while `df[:, :col]` returns a copy of the array, and changing it doesn't change `df` values.
```
Comment (bkamins) [inline docs/src/man/getting_started.md]:
```suggestion
Notice that, changing the elements of the column returned by `df[!, :col]` will affect the values stored in the original `df`, while `df[:, :col]` returns a copy of the column, and changing it does not change `df`.
```

File: docs/src/man/getting_started.md [hunk 2]
```diff
@@ -45,7 +45,9 @@ julia> df = DataFrame(A = 1:4, B = ["M", "F", "F", "M"])
 
 ```
 
-Columns can be directly (i.e. without copying) accessed via `df.col` or `df[!, :col]`. The latter syntax is more flexible as it allows passing a variable holding the name of the column, and not only a literal name. Note that column names are symbols (`:col` or `Symbol("col")`) rather than strings (`"col"`). Columns can also be accessed using an integer index specifying their position.
+Columns can be directly (i.e. without copying) accessed via `df.col` or `df[!, :col]`. The latter syntax is more flexible as it allows passing a variable holding the name of the column, and not only a literal name. Note that column names are symbols (`:col` or `Symbol("col")`) rather than strings (`"col"`). Columns can also be accessed using an integer index specifying their position. 
+
+Notice that, changing the elements of the column returned by `df[!, :col]` will affect the values stored in the original `df`, while `df[:, :col]` returns a copy of the column, and changing it does not change `df`.
```
Comment (nalimilan) [inline docs/src/man/getting_started.md]: AFAICT `df[:, :col]` hasn't been mentioned yet at this point. Can you rephrase this so that it introduces `:`, rather than making it sound like users already know it?

Comment (aminya) [reply docs/src/man/getting_started.md]: Fixed in https://github.com/JuliaData/DataFrames.jl/pull/2000/commits/c1de00de14e23f0fc19aa208fbbf5c5b130068eb

**Comment by nalimilan** [comment]
Thanks. I've made some changes, let me know what you think.

File: docs/src/man/getting_started.md [hunk 3]
```diff
@@ -47,7 +47,7 @@ julia> df = DataFrame(A = 1:4, B = ["M", "F", "F", "M"])
 
 Columns can be directly (i.e. without copying) accessed via `df.col` or `df[!, :col]`. The latter syntax is more flexible as it allows passing a variable holding the name of the column, and not only a literal name. Note that column names are symbols (`:col` or `Symbol("col")`) rather than strings (`"col"`). Columns can also be accessed using an integer index specifying their position. 
 
-Notice that, changing the elements of the column returned by `df[!, :col]` will affect the values stored in the original `df`. To get a copy of the column use `df[:, :col]`. Changing the column returned by `df[:, :col]` does not change `df`.
+Since `df[!, :col]` does not make a copy, changing the elements of the column returned by this syntax will affect the values stored in the original `df`. Use the `df[:, :col]` syntax to get a copy of the column, which can be modified without changing `df`.
```
Comment (aminya) [inline docs/src/man/getting_started.md]: This is too picky, but regarding the last part, I like my sentence structure more :smile: Combining the two sentences and using `which` makes it a little vague (like: `what can be modified?`). 
Otherwise, it looks good.

Comment (nalimilan) [reply docs/src/man/getting_started.md]: Can you suggest a wording?

Comment (aminya) [reply docs/src/man/getting_started.md]:
```suggestion
Since `df[!, :col]` does not make a copy, changing the elements of the column returned by this syntax will affect the values stored in the original `df`. 

To get a copy of the column use `df[:, :col]`. Changing the column returned by this syntax does not change `df`.
```

**Comment by scls19fr** [comment]
Maybe a good place to warn about the difference could be also in the warning

    Warning: `getindex(df::DataFrame, col_ind::ColumnIndex)` is deprecated, use `df[!, col_ind]` 

which could be changed for

    Warning: `getindex(df::DataFrame, col_ind::ColumnIndex)` is deprecated, use `df[!, col_ind]` (reference) or `df[:, col_ind]` to get a copy

when user are doing `df[:column_name]`

**Comment by aminya** [comment]
> Maybe a good place to warn about the difference could be also in the warning
> 
> ```
> Warning: `getindex(df::DataFrame, col_ind::ColumnIndex)` is deprecated, use `df[!, col_ind]` 
> ```
> 
> which could be changed for
> 
> ```
> Warning: `getindex(df::DataFrame, col_ind::ColumnIndex)` is deprecated, use `df[!, col_ind]` (reference) or `df[:, col_ind]` to get a copy
> ```
> 
> when user are doing `df[:column_name]`

`:` is not deprecated. They are two different methods.

**Comment by nalimilan** [comment]
I've slightly edited the wording.

> `:` is not deprecated.

I don't think that's what @scls19fr suggested. @scls19fr Can you file a new PR? You can edit the file directly on GitHub if you want.

**Comment by scls19fr** [comment]
Absolutely @nalimilan what I'm saying is that `df[:col]` is deprecated and `df[!, :col]` or `df[:, :col]` should be used instead (and `!` or `:` should be used depending if user want a reference or a copy of column data)

I've been searching for this depwarn using
https://github.com/JuliaData/DataFrames.jl/search?q=use+%60df%5B%21%2C+col_ind%5D%60&unscoped_q=use+%60df%5B%21%2C+col_ind%5D%60

but couldn't find where it's located.

@nalimilan could you point me to code where it's located?

I will submit a PR also

**Comment by nalimilan** [comment]
It's here: https://github.com/JuliaData/DataFrames.jl/blob/9b30a99303414839579510c4c3138752333fe446/src/deprecated.jl#L1390:L1396

**Comment by scls19fr** [comment]
Hum... not so easy to add information there I thought it could be done using a simple string... but that's not the case (or at least I don't know how to provide deprecated macro an informative string)

**Comment by nalimilan** [comment]
You have to replace this with a normal function which will call `Base.depwarn`. See examples in the file.
